### ISSUE TITLE
[3.0 Alpha] Allow pages to be editable with both Oxygen and Gutenberg

### ISSUE THREAD

**Issue by sunnyt7** [soflyy/oxygen-bugs-and-features#418]
I absolutely love the new Gutenberg integration, but I think the current workflow leaves a lot of room for improvement.

In particular, it doesn't really fit within the workflow of how most of us build our sites.

For example, let's say you're building a site where you'd like the client to easily be able to edit content with Gutenberg. The process looks like:

1. Create a temporary page like "Services - Blocks"
2. Build the page in Oxygen
3. "Copy to block" every single section.
4. Create a new page "Services"
5. Insert all the blocks in their correct order

And then if you ever want to make any significant modifications to the page, you need to make your changes in each block individually, which is tedious and breaks the flow of designing a cohesive page. This is a deal breaker IMO.

In an ideal world, pages should be editable using both Oxygen AND Gutenberg. The new workflow looks like:
1. Create Services page
2. Build the page in Oxygen

And afterwards you have the ability to edit the content in the Gutenberg editor, or "Edit in Oxygen" for any bigger layout changes.

*Note: I am not suggesting to remove the ability of adding Oxygen blocks to Gutenberg. It has a number of its own use-cases - like adding CTAs to blog posts.

**One challenge with this is 'what if a a user adds another Gutenberg block to their page?'. I'm not sure what the best solution is for this, but it would probably just need to be rendered in the Oxygen editor like any shortcode would.

**Comment by louisreingold** [comment]
What if we had a "Convert All to Blocks" button for a page?

For architectural reasons related to the way Gutenberg works (in terms of code), and how Oxygen already worked before Gutenberg, what you want will never happen, even though it would be an ideal solution.

**Comment by louisreingold** [comment]
Or if we had a "Convert to Gutenberg" option, that would convert all the sections on the page to Gutenberg blocks, place them in Gutenberg, and remove the Oxygen-designed content for that page?

**Comment by louisreingold** [comment]
Maybe we could have an "Editable in Gutenberg" checkbox that you could check on a page to enable it to be edited in Gutenberg? If so, it wouldn't possible to re-order the sections or add additional blocks. But the user could edit the content, images, icons, etc. in Gutenberg, and then you could edit the page in Oxygen. The content the user enters in would not appear in Oxygen though. Thoughts?

**Comment by sunnyt7** [comment]
#1 and #2 help reduce the number of steps, but don't solve the issue of being able to edit as a single page in Oxygen afterwards. If I can only really edit individual blocks after converting, I don't think I'd use the integration at all.

#3 on the other hand... I think that's a winner. Before reading it I was thinking I'd even prefer if the client had to click Edit in Oxygen, but got the simplified Gutenberg experience. Kind of like Elementor's role manager. Your idea might be even better though.

**Comment by sunnyt7** [comment]
The only thing is that last part about their content not showing up in Oxygen. I can see how that could become an awkward experience for some people.

Maybe just having the simplified editing experience in Oxygen is the better option? Hmm..

**Comment by sunnyt7** [comment]
I think it'll be like:

Want to create your page like usual and have the client able to easily make content changes? Great, assign them the role Editor and their Oxygen builder experience will be greatly simplified.

Want to build Gutenberg blocks using Oxygen? Great, you can do that by adding blocks to the block library. From there you can add them to any Gutenberg page, or even create a complete page using your blocks.

**Comment by wplit** [comment]
A solution to the workflow you were originally suggesting @sunnyt7..

Instead of creating the page in Oxygen first, then rebuilding in Gutenberg etc. You'd build the entire page layout as one element in your block library. Then just create a new page and drop that one block on the page.

The client would basically just be looking at one block, able to edit everything inside, unable to break the layout or accidentally delete sections as it's all one block. If you then need to add a new section in-between other sections, you'd do that from the block library. This would automatically add the section on the actual page also as it's just displaying that one block and will contain the update.

**Comment by wplit** [comment]
Testing it now. Works well this way. 

It's actually smoother experience for client, as they don't have all the Gutenberg settings appearing on each section. Assuming you don't want them to be able to reorder the sections, could be a solid solution. 

The editing they do is preserved, even as we add a new section from inside Oxygen. When they open the page editor, the new section is already there.

<img width="1300" alt="Screen Shot 2019-07-24 at 1 03 42 pm" src="https://user-images.githubusercontent.com/43051571/61761897-84f87d80-ae13-11e9-9ed2-c9f37023feda.png">

You just got to make sure you aren't saving in Oxygen at the same time they are trying to save in Gutenberg. That was the only time I lost some content.

**Comment by sunnyt7** [comment]
@wplit hah! that's actually really interesting.

**Comment by louisreingold** [comment]
That's a very good point. So really, we just need a "Convert All to Gutenberg Block" option so from a workflow standpoint you could convert an existing "normal" page to a block in the Block Library in one click.

**Comment by wplit** [comment]
Yes, but then the user would still need to create a new 'Gutenberg page' for the block to exist in, because the current page that they are "Convertting All to Gutenberg Block" will be a page edited by Oxygen & not a Gutenberg one. So you still have doubling up on pages, unless you make it that once they have clicked the convert button, that page no longer acts as an Oxygen page and is only displaying blocks now. So after converting, the next thing they would do is add their newly created block back into that same page.

As sunny mentioned, there are a few use cases for wanting to use blocks & Oxygen. So it could get confusing as not everybody will be using it the same way. Three I can think of..

1. The user wants to allow clients to edit their complete Oxygen-designed pages, visually inside the editor. This is what we're discussing now.
2. The user wants to just create extra blocks for their post content without having to resort to using Gutenberg add-on packs (for in-post call to actions, pricing tables etc)
3. The user wants to allow clients to be able to build pages themselves using blocks (seperate sections that have been built in Oxygen) and change the order of the blocks and add their own blocks that aren't from Oxygen on the same page. Essentially giving them a simple block page builder to use. (for allowing editors to construct landing pages / sales pages where they are likely to want to move around content) 

There's probably more.

**Comment by wplit** [comment]
I'm probably looking in to it too deep, a "Convert All to Gutenberg Block" would work pretty well.

**Comment by sunnyt7** [comment]
When it comes to use case #1 - clients editing content - what is the advantage of having this done in the Gutenberg editor as opposed to a simplified role-based Oxygen editor experience?

Only things I can think of are not having to wait for the Oxygen editor to load, and possible Gutenberg familiarity.

Is that worth having to convert [legacy] pages into blocks? Or explaining to a new Oxygen user that they can make their pages as a normal page or can create it as a block and then insert into a page?

I like where this is heading. Still just needs a bit of refinement I think.

**Comment by sunnyt7** [comment]
This is rough but just thinking out loud.

What if when you go to create a new page, there's a checkbox in the Oxygen meta box that says "Edit With Gutenberg Mode".

When you check that box and hit update, it converts the page to a block, automatically inserts it into the Gutenberg editor (if possible), and Edit With Oxygen becomes something like Edit Block in Oxygen.

**Comment by Deanphillips** [comment]
The more I test the GB add on, the more I really hope this suggestion + https://github.com/soflyy/oxygen-bugs-and-features/issues/393 happen. It would really improve this whole GB feature set.

**Comment by Deanphillips** [comment]
The main annoyance for me is creating a page all in O2, then making it a block (or blocks) then creating a new page, changing URL slugs on both and adding the blocks to a blank page. So many steps to convert a page. 

Then there's the issue when elements aren't being picked up and not being able to select what's changable in the Gutenburg block itself (to both give more control and make sure O2 picks up all the parts that should be editable)

**Comment by rytly** [comment]
+1
